File: src/Cxx/PolyData/ThinPlateSplineTransform.cxx
#include <vtkImageGridSource.h>
#include <vtkLookupTable.h>
#include <vtkImageMapToColors.h>
#include <vtkBMPReader.h>
#include <vtkImageBlend.h>
#include <vtkPoints.h>
#include <vtkThinPlateSplineTransform.h>
#include <vtkImageReslice.h>
#include <vtkSmartPointer.h>
#include <vtkDataObject.h>
#include <vtkRenderer.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkInteractorStyleTerrain.h>
#include <vtkImageMapper.h>
#include <vtkActor2D.h>

// Warp an image with a thin plate spline

int main(int argc, char *argv[])
{
  if ( argc < 2 )
  {
    std::cerr << "Usage: " << argv[0] << " Filename.bmp" << std::endl;
    return EXIT_FAILURE;
  }

  // First, create an image to warp
  vtkSmartPointer<vtkImageGridSource> imageGrid =
    vtkSmartPointer<vtkImageGridSource >::New();
  imageGrid->SetGridSpacing( 16, 16, 0 );
  imageGrid->SetGridOrigin( 0, 0, 0 );
  imageGrid->SetDataExtent( 0, 255, 0, 255, 0, 0 );
  imageGrid->SetDataScalarTypeToUnsignedChar();

  vtkSmartPointer<vtkLookupTable> table =
    vtkSmartPointer<vtkLookupTable >::New();
  table->SetTableRange( 0, 1 );
  table->SetAlphaRange( 0.0, 1.0 );
  table->SetHueRange( 0.15, 0.15 );
  table->SetSaturationRange( 1, 1 );
  table->SetValueRange( 0, 1 );
  table->Build();

  vtkSmartPointer<vtkImageMapToColors> alpha =
    vtkSmartPointer<vtkImageMapToColors >::New();
  alpha->SetInputConnection( imageGrid->GetOutputPort() );
  alpha->SetLookupTable( table );

  vtkSmartPointer<vtkBMPReader> reader =
    vtkSmartPointer<vtkBMPReader>::New();
  reader->SetFileName(argv[1]);
  reader->Update();

  vtkSmartPointer<vtkImageBlend> blend =
    vtkSmartPointer<vtkImageBlend >::New();
  blend->AddInputConnection( 0, reader->GetOutputPort() );
  blend->AddInputConnection( 0, alpha->GetOutputPort() );

  // Next, create a ThinPlateSpline transform

  vtkSmartPointer< vtkPoints > p1 = vtkSmartPointer< vtkPoints >::New();
  p1->SetNumberOfPoints( 8 );
  p1->SetPoint( 0, 0, 0, 0 );
  p1->SetPoint( 1, 0, 255, 0);
  p1->SetPoint( 2, 255, 0, 0 );
  p1->SetPoint( 3, 255, 255, 0 );
  p1->SetPoint( 4, 96, 96, 0 );
  p1->SetPoint( 5, 96, 159, 0 );
  p1->SetPoint( 6, 159, 159, 0 );
  p1->SetPoint( 7, 159, 96, 0 );

  vtkSmartPointer< vtkPoints > p2 = vtkSmartPointer< vtkPoints >::New();
  p2->SetNumberOfPoints( 8 );
  p2->SetPoint( 0, 0, 0, 0 );
  p2->SetPoint( 1, 0, 255, 0 );
  p2->SetPoint( 2, 255, 0, 0 );
  p2->SetPoint( 3, 255, 255, 0);
  p2->SetPoint( 4, 96, 159, 0 );
  p2->SetPoint( 5, 159, 159, 0 );
  p2->SetPoint( 6, 159, 96, 0 );
  p2->SetPoint( 7, 96, 96, 0 );

  vtkSmartPointer<vtkThinPlateSplineTransform> transform =
    vtkSmartPointer< vtkThinPlateSplineTransform >::New();
  transform->SetSourceLandmarks( p2 );
  transform->SetTargetLandmarks( p1 );
  transform->SetBasisToR2LogR();
  // You must invert the transform before passing it to vtkImageReslice
  transform->Inverse();

  vtkSmartPointer< vtkImageReslice > reslice =
    vtkSmartPointer<vtkImageReslice >::New();
  reslice->SetInputConnection( blend->GetOutputPort() );
  reslice->SetResliceTransform( transform );
  reslice->SetInterpolationModeToLinear();
  vtkSmartPointer< vtkImageMapper > map =
    vtkSmartPointer<vtkImageMapper >::New();
  map->SetInputConnection( reslice->GetOutputPort() );
  map->SetColorWindow( 255.0);
  map->SetColorLevel( 127.5 );
  map->SetZSlice( 0 );
  vtkSmartPointer< vtkActor2D > act =
    vtkSmartPointer< vtkActor2D >::New();
  act->SetMapper( map );
  act->SetPosition( 0.0, 0.0 );

  vtkSmartPointer< vtkRenderer > renderer =
    vtkSmartPointer<vtkRenderer >::New();
  renderer->AddActor( act );
  vtkSmartPointer< vtkRenderWindow > window =
    vtkSmartPointer<vtkRenderWindow >::New();
  window->SetSize( 640, 480 );
  window->AddRenderer(renderer);
  vtkSmartPointer< vtkRenderWindowInteractor > interactor =
    vtkSmartPointer< vtkRenderWindowInteractor >::New();
  interactor->SetRenderWindow(window);
  vtkSmartPointer< vtkInteractorStyleTerrain > style =
    vtkSmartPointer< vtkInteractorStyleTerrain >::New();
  interactor->SetInteractorStyle( style );
  window->Render();
  interactor->Initialize();
  interactor->Start();

  return EXIT_SUCCESS;
}

File: src/Java/PolyData/ThinPlateSplineTransform.java
import vtk.vtkNamedColors;
import vtk.vtkNativeLibrary;
import vtk.vtkRenderWindow;
import vtk.vtkRenderWindowInteractor;
import vtk.vtkRenderer;
import vtk.vtkImageGridSource;
import vtk.vtkLookupTable;
import vtk.vtkImageMapToColors;
import vtk.vtkBMPReader;
import vtk.vtkImageBlend;
import vtk.vtkPoints;
import vtk.vtkThinPlateSplineTransform;
import vtk.vtkImageReslice;
import vtk.vtkImageMapper;
import vtk.vtkActor2D;
import vtk.vtkInteractorStyleTerrain;

public class ThinPlateSplineTransform 
{
  // -----------------------------------------------------------------
  // Load VTK library and print which library was not properly loaded
  static 
  {
    if (!vtkNativeLibrary.LoadAllNativeLibraries()) 
    {
      for (vtkNativeLibrary lib : vtkNativeLibrary.values()) 
      {
        if (!lib.IsLoaded()) 
        {
          System.out.println(lib.GetLibraryName() + " not loaded");
        }
      }
    }
    vtkNativeLibrary.DisableOutputWindow(null);
  }
  // -----------------------------------------------------------------

  public static void main(String args[]) 
  {
    //parse command line arguments
    if (args.length != 1) 
    {
      System.err.println("Usage: java -classpath ... Filename(.bmp) e.g masonry.bmp");
      return;
    }
    String inputFilename = args[0];
    
    vtkNamedColors colors = new vtkNamedColors();

    //Renderer Background Color
    double Bgcolor[] = new double[4];

    colors.GetColor("Black", Bgcolor);
    
    // First, create an image to warp
    vtkImageGridSource imageGrid = new vtkImageGridSource();
    imageGrid.SetGridSpacing(16, 16, 0);
    imageGrid.SetGridOrigin(0, 0, 0);
    imageGrid.SetDataExtent(0, 255, 0, 255, 0, 0);
    imageGrid.SetDataScalarTypeToUnsignedChar();

    vtkLookupTable table = new vtkLookupTable();
    table.SetTableRange(0, 1);
    table.SetAlphaRange(0.0, 1.0);
    table.SetHueRange(0.15, 0.15);
    table.SetSaturationRange(1, 1);
    table.SetValueRange(0, 1 );
    table.Build();
    
    vtkImageMapToColors alpha = new vtkImageMapToColors();
    alpha.SetInputConnection(imageGrid.GetOutputPort());
    alpha.SetLookupTable( table );
    
    vtkBMPReader reader = new vtkBMPReader();
    reader.SetFileName(inputFilename);
    reader.Update();
    
    vtkImageBlend blend = new vtkImageBlend();
    blend.AddInputConnection(0, reader.GetOutputPort());
    blend.AddInputConnection(0, alpha.GetOutputPort());
    
    // Next, create a ThinPlateSpline transform
    vtkPoints p1 = new vtkPoints();
    p1.SetNumberOfPoints( 8 );
    p1.SetPoint(0, 0, 0, 0);
    p1.SetPoint(1, 0, 255, 0);
    p1.SetPoint(2, 255, 0, 0);
    p1.SetPoint(3, 255, 255, 0);
    p1.SetPoint(4, 96, 96, 0);
    p1.SetPoint(5, 96, 159, 0);
    p1.SetPoint(6, 159, 159, 0);
    p1.SetPoint(7, 159, 96, 0);

    vtkPoints p2 = new vtkPoints();
    p2.SetNumberOfPoints( 8 );
    p2.SetPoint(0, 0, 0, 0);
    p2.SetPoint(1, 0, 255, 0);
    p2.SetPoint(2, 255, 0, 0);
    p2.SetPoint(3, 255, 255, 0);
    p2.SetPoint(4, 96, 159, 0);
    p2.SetPoint(5, 159, 159, 0);
    p2.SetPoint(6, 159, 96, 0);
    p2.SetPoint(7, 96, 96, 0);
    
    vtkThinPlateSplineTransform transform = new vtkThinPlateSplineTransform();
    transform.SetSourceLandmarks(p2);
    transform.SetTargetLandmarks(p1);
    transform.SetBasisToR2LogR();
    // You must invert the transform before passing it to vtkImageReslice
    transform.Inverse();

    vtkImageReslice reslice = new vtkImageReslice();
    reslice.SetInputConnection(blend.GetOutputPort());
    reslice.SetResliceTransform(transform);
    reslice.SetInterpolationModeToLinear();
    
    vtkImageMapper map = new vtkImageMapper();
    map.SetInputConnection(reslice.GetOutputPort());
    map.SetColorWindow( 255.0);
    map.SetColorLevel(127.5);
    map.SetZSlice(0);
    
    vtkActor2D  actor = new vtkActor2D();
    actor.SetMapper(map);
    actor.SetPosition(0.0, 0.0);

    // Create the renderer, render window and interactor.
    vtkRenderer ren = new vtkRenderer();
    vtkRenderWindow renWin = new vtkRenderWindow();
    renWin.AddRenderer(ren);
    vtkRenderWindowInteractor iren = new vtkRenderWindowInteractor();
    iren.SetRenderWindow(renWin);
    
    vtkInteractorStyleTerrain style = new vtkInteractorStyleTerrain();
    iren.SetInteractorStyle(style);
    
    // Visualize
    ren.AddActor(actor);
	
    ren.SetBackground(Bgcolor);
    
    renWin.SetSize(640, 480);
    renWin.Render();

    iren.Initialize();
    iren.Start();
  }
}
